lib/auth/oidc.go

```diff
@@ -649,6 +662,10 @@ func (a *Server) getClaims(oidcClient *oidc.Client, connector types.OIDCConnecto
 			log.Debugf("OIDC provider doesn't offer valid UserInfo endpoint. Returning token claims: %v.", idTokenClaims)
 			return idTokenClaims, nil
 		}
+		if trace.IsAccessDenied(err) {
```

Comment:
Should we specifically look for 401?

Unless I'm reading it wrong, It looks like AccessDenied includes anything that isn't a 2xx in this case ([lib/auth/oidc.go#595](https://github.com/gravitational/teleport/blob/2d958d69571f5c028ceb51e55ce9d01e9c520f2c/lib/auth/oidc.go#L595-L597)).

Reply:
Right. According to [the spec](https://openid.net/specs/openid-connect-core-1_0.html#UserInfoError) and [RFC 6750](https://datatracker.ietf.org/doc/html/rfc6750#section-3.1), UserInfo can return any appropriate status code. I don't think we need to filter for 401 unless we want to take different actions for different codes.

Reply:
You have much better context here than me, so let me know if this doesn't make sense:

I think I'd rather check only for the handful of statuses that make sense/we know happen (400, 401, 403, or 405, according to the RFC), but still fail if we get something outside of that range, so if something too weird happens it won't be silent. If this comes and bites us again, then I would relax the check.

In any case, please add a comment in the code calling to attention the fact that are handling a various of statuses with this if; usually I would expect AccessDenied to be 401/403.

Reply:
That sounds fine. Is it worth adding custom errors in [trace](https://github.com/gravitational/trace) for 401 and 405, or should I stick with bundling them all under `AccessDenied`?

Reply:
No need to hold your PR on a trace change, I think. Either bundle in AccessDenied or return specific errors for different types, up to you - just mention in `claimsFromUserInfo`'s docs what AccessDenied means there, if you go that route.